### ISSUE TITLE
Unable to relock bootloader

### ISSUE THREAD

**Issue by kcespedes** [chenxiaolong/avbroot#305]
I followed the guide successfully until last step. However, when I attempt re-locking the bootloader there is the following error message:

% fastboot flashing lock

FAILED (remote: 'invalid android images, skip locking')

I followed these steps:

1 downloaded OTA image and stock image from the google website for Pixel 6 pro
2 created the necessary keys
3 created a patched image with magisk of the stock image boot.img
4 patched the OTA.zip file with the signing keys created and the magisk patched stock image
5 applied the patched OTA
6 added custom avb keys to bootloader
7 checked with this command: adb shell su -c 'dmesg | grep libfs_avb'
and did receive the following response: init: [libfs_avb]Returning avb_handle with status: Success

what am I missing if all steps completed successfully? Why am I not able to relock the bootloader?

**Comment by chenxiaolong** [comment]
> 5 applied the patched OTA
> 6 added custom avb keys to bootloader

Could you try these again and post the output of the commands here?

I'm not sure if it's the case here, but `invalid android images, skip locking` usually means either `avb_custom_key` or `vbmeta` did not get flashed properly.

**Comment by derzahla** [comment]
ugh, im getting this too on a new pixel 7 while trying to install a custom key version of grapheneos. 

I tried flashing the factory AOSP image, I am able to relock bootloader.
Tried flashing GrapheneOS factory image, I am able to relock the bootloader.

But as soon I erase avb_custom_key and my custom key: fastboot flash avb_custom_key priv/avb_pkmd.bin
I get the same error as OP when trying to  relock the bootloader.

I have also tried replacing the avb_pkmd.bin with my own in the extracted grapheneos factory image and running flash-all.sh - which it completes successfully I am still not able to relocked my bootloader.  As could probably be expected, patching the latest grapheneos OTA with avbroot and my custom key fails on signature verification when attempting to sideload.  I have even tried regenerating fresh keys/certs with avbroot - same error.  This was working just fine on my pixel 5, any ideas what else I can try?  Thanks!

`panther-factory-2024072800]$ ./flash-all.sh 
Warning: skip copying bootloader_b image avb footer (bootloader_b partition size: 0, bootloader_b image size: 14655240).
Sending 'bootloader_b' (14311 KB)                  OKAY [  0.335s]
Writing 'bootloader_b'                             (bootloader) Flashing pack version cloudripper-14.5-11677884
(bootloader) flashing platform gs201
(bootloader) Validating partition ufs
(bootloader) Validating partition ufs
(bootloader) Validating partition partition:0
(bootloader) Validating partition partition:1
(bootloader) Validating partition partition:2
(bootloader) Validating partition partition:3
(bootloader) Validating partition bl1_b
(bootloader) Validating partition pbl_b
(bootloader) Validating partition bl2_b
(bootloader) Validating partition abl_b
(bootloader) Validating partition bl31_b
(bootloader) Validating partition tzsw_b
(bootloader) Validating partition gsa_b
(bootloader) Validating partition ldfw_b
(bootloader) Flashing partition ufs
(bootloader) Flashing partition ufs
(bootloader) Flashing partition partition:0
(bootloader) Flashing partition partition:1
(bootloader) Flashing partition partition:2
(bootloader) Flashing partition partition:3
(bootloader) Flashing partition bl1_b
(bootloader) Flashing partition pbl_b
(bootloader) Flashing partition bl2_b
(bootloader) Flashing partition abl_b
(bootloader) Flashing partition bl31_b
(bootloader) Flashing partition tzsw_b
(bootloader) Flashing partition gsa_b
(bootloader) Flashing partition ldfw_b
(bootloader) Loading sideload ufsfwupdate
OKAY [  0.189s]
Finished. Total time: 0.526s
Setting current slot to 'b'                        OKAY [  0.064s]
Finished. Total time: 0.065s
Rebooting into bootloader                          OKAY [  0.000s]
Finished. Total time: 0.050s
< waiting for any device >
Warning: skip copying bootloader_a image avb footer (bootloader_a partition size: 0, bootloader_a image size: 14655240).
Sending 'bootloader_a' (14311 KB)                  OKAY [  0.336s]
Writing 'bootloader_a'                             (bootloader) Flashing pack version cloudripper-14.5-11677884
(bootloader) flashing platform gs201
(bootloader) Validating partition ufs
(bootloader) Validating partition ufs
(bootloader) Validating partition partition:0
(bootloader) Validating partition partition:1
(bootloader) Validating partition partition:2
(bootloader) Validating partition partition:3
(bootloader) Validating partition bl1_a
(bootloader) Validating partition pbl_a
(bootloader) Validating partition bl2_a
(bootloader) Validating partition abl_a
(bootloader) Validating partition bl31_a
(bootloader) Validating partition tzsw_a
(bootloader) Validating partition gsa_a
(bootloader) Validating partition ldfw_a
(bootloader) Flashing partition ufs
(bootloader) Flashing partition ufs
(bootloader) Flashing partition partition:0
(bootloader) Flashing partition partition:1
(bootloader) Flashing partition partition:2
(bootloader) Flashing partition partition:3
(bootloader) Flashing partition bl1_a
(bootloader) Flashing partition pbl_a
(bootloader) Flashing partition bl2_a
(bootloader) Flashing partition abl_a
(bootloader) Flashing partition bl31_a
(bootloader) Flashing partition tzsw_a
(bootloader) Flashing partition gsa_a
(bootloader) Flashing partition ldfw_a
(bootloader) Loading sideload ufsfwupdate
OKAY [  0.157s]
Finished. Total time: 0.512s
Setting current slot to 'a'                        OKAY [  0.072s]
Finished. Total time: 0.073s
Rebooting into bootloader                          OKAY [  0.000s]
Finished. Total time: 0.050s
< waiting for any device >
Warning: skip copying radio_a image avb footer (radio_a partition size: 0, radio_a image size: 82227340).
Sending 'radio_a' (80300 KB)                       OKAY [  1.883s]
Writing 'radio_a'                                  (bootloader) Flashing pack version g5300q-240308-240517-M-11857457
(bootloader) Flashing partition modem_a
OKAY [  0.157s]
Finished. Total time: 2.060s
Rebooting into bootloader                          OKAY [  0.000s]
Finished. Total time: 0.050s
< waiting for any device >
Erasing 'avb_custom_key'                           (bootloader) avb custom key: erase done
OKAY [  0.008s]
Finished. Total time: 0.009s
Warning: skip copying avb_custom_key image avb footer (avb_custom_key partition size: 0, avb_custom_key image size: 1032).
Sending 'avb_custom_key' (1 KB)                    OKAY [  0.000s]
Writing 'avb_custom_key'                           (bootloader) avb custom key: flash done
OKAY [  0.029s]
Finished. Total time: 0.031s
OKAY [  0.005s]
Finished. Total time: 0.005s
Erasing 'fips'                                     OKAY [  0.027s]
Finished. Total time: 0.028s
Erasing 'dpm_a'                                    OKAY [  0.014s]
Finished. Total time: 0.014s
Erasing 'dpm_b'                                    OKAY [  0.009s]
Finished. Total time: 0.009s
Snapshot cancel                                    OKAY [  0.002s]
Finished. Total time: 0.002s
--------------------------------------------
Bootloader Version...: cloudripper-14.5-11677884
Baseband Version.....: g5300q-240308-240517-B-11857457
Serial Number........: 27281FDH200BSU
--------------------------------------------
extracting android-info.txt (0 MB) to RAM...
Checking 'product'                                 OKAY [  0.000s]
Checking 'version-bootloader'                      OKAY [  0.000s]
Checking 'version-baseband'                        OKAY [  0.000s]
Setting current slot to 'a'                        OKAY [  0.058s]
extracting fastboot-info.txt (0 MB) to RAM...
extracting super_empty.img (0 MB) to RAM...
extracting super_empty.img (0 MB) to disk... took 0.000s
extracting system.img (1195 MB) to disk... took 3.838s
[liblp] Partition system_a will resize from 0 bytes to 1253756928 bytes
extracting system_dlkm.img (0 MB) to disk... took 0.001s
[liblp] Partition system_dlkm_a will resize from 0 bytes to 348160 bytes
extracting system_ext.img (378 MB) to disk... took 1.121s
[liblp] Partition system_ext_a will resize from 0 bytes to 396537856 bytes
extracting product.img (676 MB) to disk... took 2.111s
[liblp] Partition product_a will resize from 0 bytes to 709820416 bytes
extracting vendor.img (607 MB) to disk... took 1.676s
[liblp] Partition vendor_a will resize from 0 bytes to 636919808 bytes
extracting vendor_dlkm.img (43 MB) to disk... took 0.088s
[liblp] Partition vendor_dlkm_a will resize from 0 bytes to 45412352 bytes
extracting super_empty.img (0 MB) to RAM...
extracting boot.img (64 MB) to disk... took 0.175s
archive does not contain 'boot.sig'
extracting super_empty.img (0 MB) to RAM...
Sending 'boot_a' (65536 KB)                        OKAY [  1.539s]
Writing 'boot_a'                                   OKAY [  0.125s]
extracting super_empty.img (0 MB) to RAM...
extracting init_boot.img (8 MB) to disk... took 0.051s
archive does not contain 'init_boot.sig'
extracting super_empty.img (0 MB) to RAM...
Sending 'init_boot_a' (8192 KB)                    OKAY [  0.193s]
Writing 'init_boot_a'                              OKAY [  0.037s]
extracting super_empty.img (0 MB) to RAM...
extracting dtbo.img (16 MB) to disk... took 0.081s
archive does not contain 'dtbo.sig'
extracting super_empty.img (0 MB) to RAM...
Sending 'dtbo_a' (16384 KB)                        OKAY [  0.384s]
Writing 'dtbo_a'                                   OKAY [  0.045s]
extracting super_empty.img (0 MB) to RAM...
extracting vendor_kernel_boot.img (64 MB) to disk... took 0.186s
archive does not contain 'vendor_kernel_boot.sig'
extracting super_empty.img (0 MB) to RAM...
Sending 'vendor_kernel_boot_a' (65536 KB)          OKAY [  1.535s]
Writing 'vendor_kernel_boot_a'                     OKAY [  0.120s]
extracting super_empty.img (0 MB) to RAM...
extracting pvmfw.img (1 MB) to disk... took 0.013s
archive does not contain 'pvmfw.sig'
extracting super_empty.img (0 MB) to RAM...
Sending 'pvmfw_a' (1024 KB)                        OKAY [  0.025s]
Writing 'pvmfw_a'                                  OKAY [  0.024s]
extracting super_empty.img (0 MB) to RAM...
extracting vendor_boot.img (64 MB) to disk... took 0.225s
archive does not contain 'vendor_boot.sig'
extracting super_empty.img (0 MB) to RAM...
Sending 'vendor_boot_a' (65536 KB)                 OKAY [  1.533s]
Writing 'vendor_boot_a'                            OKAY [  0.117s]
extracting super_empty.img (0 MB) to RAM...
extracting vbmeta.img (0 MB) to disk... took 0.000s
archive does not contain 'vbmeta.sig'
extracting super_empty.img (0 MB) to RAM...
Sending 'vbmeta_a' (8 KB)                          OKAY [  0.001s]
Writing 'vbmeta_a'                                 OKAY [  0.020s]
Erasing 'userdata'                                 OKAY [  0.749s]
Erase successful, but not automatically formatting.
File system type raw not supported.
Erasing 'metadata'                                 OKAY [  0.014s]
Erase successful, but not automatically formatting.
File system type raw not supported.
Sending sparse 'super' 1/12 (254972 KB)            OKAY [  5.983s]
Writing 'super'                                    OKAY [  0.372s]
Sending sparse 'super' 2/12 (254972 KB)            OKAY [  5.982s]
Writing 'super'                                    OKAY [  0.394s]
Sending sparse 'super' 3/12 (254972 KB)            OKAY [  5.975s]
Writing 'super'                                    OKAY [  0.394s]
Sending sparse 'super' 4/12 (254972 KB)            OKAY [  5.970s]
Writing 'super'                                    OKAY [  0.432s]
Sending sparse 'super' 5/12 (254972 KB)            OKAY [  5.977s]
Writing 'super'                                    OKAY [  0.450s]
Sending sparse 'super' 6/12 (254972 KB)            OKAY [  5.978s]
Writing 'super'                                    OKAY [  0.425s]
Sending sparse 'super' 7/12 (254972 KB)            OKAY [  5.992s]
Writing 'super'                                    OKAY [  0.430s]
Sending sparse 'super' 8/12 (254972 KB)            OKAY [  6.001s]
Writing 'super'                                    OKAY [  0.428s]
Sending sparse 'super' 9/12 (254972 KB)            OKAY [  5.998s]
Writing 'super'                                    OKAY [  0.425s]
Sending sparse 'super' 10/12 (254972 KB)           OKAY [  5.998s]
Writing 'super'                                    OKAY [  0.428s]
Sending sparse 'super' 11/12 (254972 KB)           OKAY [  5.994s]
Writing 'super'                                    OKAY [  0.426s]
Sending sparse 'super' 12/12 (166820 KB)           OKAY [  3.924s]
Writing 'super'                                    OKAY [  0.294s]
Erasing 'userdata'                                 OKAY [  0.724s]
Erase successful, but not automatically formatting.
File system type raw not supported.
wipe task partition not found: cache
Erasing 'metadata'                                 OKAY [  0.014s]
Erase successful, but not automatically formatting.
File system type raw not supported.
Finished. Total time: 91.716s
Rebooting into bootloader                          OKAY [  0.000s]
Finished. Total time: 0.050s
[jzr3@tyr panther-factory-2024072800]$ fastboot flashing lock
< waiting for any device >
FAILED (remote: 'invalid android images, skip locking')
fastboot: error: Command failed
`

**Comment by chenxiaolong** [comment]
Can you try flashing your patched OTA (as you've already done) and the sideload that same OTA from recovery twice before relocking the bootloader?

This will ensure that both the A and B slots are running your custom signed OS. I'm curious if that would fix the problem.

**Comment by derzahla** [comment]
My bad,  it's been awhile since I had to start over with a stock phone.  I forgot that I needed to extract the patched OTA and then flash each partition from the extracted patched OTA manually BEFORE I could sideload the patched OTA successfully.  I was thinking that just by extracting the factory image, replacing avb_pkmd.bin and running flash_all.sh that my key was being installed - but I guess not.  It is working as expected now.  Thanks!!
